### PR TITLE
vttablet mysql conn improvements

### PR DESCRIPTION
**Pull request by demmer**
Add a couple of changes to improve visibility and robustness on the vttablet => mysql connection handling.

1. Add timing stats for successful / failed calls to Connect to mysql.

In order to better monitor how many times the tablet creates a new mysql connection and how long these attempts take, add another call to measure the timings of each Connect operation in the dbconn wrapper.

Additionally record the timings and associated counts for any times
when the connection to mysql fails to execute the handshake.

2. Check that the context deadline hasn't expired before running a query

If the context deadline expires before the query has a chance to execute, the query killer will terminate the connection immediately after sending the query to mysql.

Instead, add a check at the start of dbConn.execOnce and fail the query immediately if the context is already past the deadline.

### CODE REVIEW

**Comment by demmer** [comment]
NB: I added simple unit tests but haven't yet tested this in a real environment. Will do that tomorrow.

File: go/vt/vttablet/tabletserver/connpool/dbconn.go [hunk 1]
```diff
@@ -132,6 +132,16 @@ func (dbc *DBConn) execOnce(ctx context.Context, query string, maxrows int, want
 	dbc.current.Set(query)
 	defer dbc.current.Set("")
 
+	// Check if the context is already past its deadline before
+	// trying to execute the query.
+	if ctx.Done() != nil {
```
Comment (sougou) [inline go/vt/vttablet/tabletserver/connpool/dbconn.go]: This check is not needed. If the channel is nil, it won't fire.

**Comment by demmer** [comment]
Oh. I copied that from other parts of the code. 

-m

File: go/vt/vttablet/tabletserver/connpool/dbconn.go (hunk 1 above)
Comment (demmer) [reply go/vt/vttablet/tabletserver/connpool/dbconn.go]: Thanks -- removed (and verified it still works)

**Pull request merged**